src/StreamJsonRpc/JsonRpcEventSource.cs

```diff
@@ -0,0 +1,232 @@
+        }
+
+        /// <summary>
+        /// Invoke Method end.
+        /// </summary>
+        /// <param name="message">Request and method details.</param>
+        [Event(InvokeMethodStopEvent, Level = EventLevel.Verbose)]
+        public void InvokeMethodStop(string message)
+        {
+            this.WriteEvent(InvokeMethodStopEvent, message);
+        }
+
+        /// <summary>
+        /// Cancel Request start.
+        /// </summary>
+        /// <param name="message">Cancellation message.</param>
+        [Event(CancelRequestStartEvent, Level = EventLevel.Verbose)]
+        public void CancelRequestStart(string message)
+        {
+            this.WriteEvent(CancelRequestStartEvent, message);
+        }
+
+        /// <summary>
+        /// Cancel Request end.
+        /// </summary>
+        /// <param name="message">Cancellation message.</param>
+        [Event(CancelRequestStopEvent, Level = EventLevel.Verbose)]
+        public void CancelRequestStop(string message)
+        {
+            this.WriteEvent(CancelRequestStopEvent, message);
+        }
+
+        /// <summary>
+        /// Invoke Notification.
+        /// </summary>
+        /// <param name="message">Details of the Notification.</param>
+        [Event(InvokeNotificationStartEvent, Level = EventLevel.Verbose)]
+        public void InvokeNotificationStart(string message)
+        {
+            this.WriteEvent(InvokeNotificationStartEvent, message);
+        }
+
+        /// <summary>
+        /// Notification completed.
+        /// </summary>
+        /// <param name="message">Details of the Notification.</param>
+        [Event(InvokeNotificationStopEvent, Level = EventLevel.Verbose)]
+        public void InvokeNotificationStop(string message)
+        {
+            this.WriteEvent(InvokeNotificationStopEvent, message);
+        }
+
+        /// <summary>
+        /// Creates a string representation of arguments of max 1024 characters for logging.
+        /// </summary>
+        /// <param name="arguments">A single argument or an array of arguments.</param>
+        /// <returns>String representation of first argument only.</returns>
+        internal static string GetArgumentsString(object arguments)
+        {
+            if (arguments == null)
+            {
+                return string.Empty;
+            }
+
+            const int maxLength = 1024;
+
+            if (arguments is object[] args)
+            {
+                var stringBuilder = new StringBuilder();
+                if (args != null && args.Length > 0)
+                {
+                    for (int i = 0; i < args.Length; ++i)
+                    {
+                        stringBuilder.Append($"arg{i}: {args[i].ToString()}, ");
+                        if (stringBuilder.Length > maxLength)
+                        {
+                            return $"{stringBuilder.ToString().Substring(0, maxLength)}...(truncated)";
```

Comment:
Your `.ToString().Substring()` pattern would end up creating the very large string only to chop it, and all after you've already constructed a long stringbuilder.
You can optimize this by replacing `.ToString().Substring()` with just `.ToString(0, maxLength)`.

Reply:
Thanks. Will do.